aptos-move/framework/aptos-framework/sources/voting.move

```diff
@@ -326,6 +373,67 @@ module aptos_framework::voting {
+    /// Can only be done if there are at least as many votes as min required and
+    /// there are more yes votes than no. If either of these conditions is not met, this will revert.
+    ///
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param next_execution_hash The next execution hash if the given proposal is multi-step.
+    public fun resolve_single_step_or_multi_step_proposal<ProposalType: store>(
+        voting_forum_address: address,
+        proposal_id: u64,
+        next_execution_hash: vector<u8>,
+    ) acquires VotingForum {
+        let proposal_state = get_proposal_state<ProposalType>(voting_forum_address, proposal_id);
+        assert!(proposal_state == PROPOSAL_STATE_SUCCEEDED, error::invalid_state(EPROPOSAL_CANNOT_BE_RESOLVED));
+
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+
+        // If this proposal is multi-step, we update the metadata map to indicate that this proposal is in execution.
+        let multi_step_in_execution_key = utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        if (simple_map::contains_key(&proposal.metadata, &multi_step_in_execution_key)) {
+            let is_multi_step_proposal_in_execution_value = simple_map::borrow_mut(&mut proposal.metadata, &multi_step_in_execution_key);
+            if (!from_bcs::to_bool(*is_multi_step_proposal_in_execution_value)) {
+                *is_multi_step_proposal_in_execution_value = to_bytes(&true);
+            };
+        };
+
+        let resolvable_time = to_u64(*simple_map::borrow(&proposal.metadata, &utf8(RESOLVABLE_TIME_METADATA_KEY)));
+        assert!(timestamp::now_seconds() > resolvable_time, error::invalid_state(ERESOLUTION_CANNOT_BE_ATOMIC));
+
+        assert!(
+            transaction_context::get_script_hash() == proposal.execution_hash,
+            error::invalid_argument(EPROPOSAL_EXECUTION_HASH_NOT_MATCHING),
+        );
+
+        let multi_step_key = utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        let is_multi_step = simple_map::contains_key(&proposal.metadata, &multi_step_key) && *simple_map::borrow(&proposal.metadata, &multi_step_key) == to_bytes(&true);
+        // If the proposal is a single-step proposal, or if this step is the last step in a multi-step proposal,
+        // we can mark that this proposal is resolved and emit a `ResolveProposal` event.
+        if (!is_multi_step || vector::length(&next_execution_hash) == 0) {
+            let resolved_early = can_be_resolved_early(proposal);
+            proposal.is_resolved = true;
+            proposal.resolution_time_secs = timestamp::now_seconds();
+
+            event::emit_event<ResolveProposal>(
```

Comment:
As discussed in the past, what do you think about emitting this event for every step in the multi-step execution? Without introducing a new event, we're not really tracking executing each step.

Reply:
I thought we didn't want to add a new event because we'll have to do two upgrades (with the second one initializing the event struct)? Do we want to add a new `ResolveMultiStepProposal` event and where do we put that event (we cannot put that in `VotingForum`?) 

Yeah I think emitting this event for every step is simpler, just not sure if it'd be confusing to people. I was thinking that we can track each step by looking up transactions on-chain by I guess that's harder / more hassle.

Reply:
Yeah ideally we would have a different event, but emitting this same event might still work? Downstream systems just need to deal with multiple resolve events for the same proposal.

Reply:
ok sounds good, updating now : )